src/Elmish/Hooks/UseEffect.purs

```diff
@@ -26,24 +28,68 @@ import Elmish.Hooks.Type (Hook, mkHook, uniqueNameFromCurrentCallStack, uniqueNa
 -- |   pure $ H.fragment $ todoView <$> todos
 -- | ```
 useEffect :: Aff Unit -> Hook Unit
-useEffect = useEffect' identity uniqueNameFromCurrentCallStack
+useEffect aff =
+  useEffect_ name unit $ const aff
+  where
+    name = uniqueNameFromCurrentCallStack { skipFrames: 3 }
+
+-- | This is like `useEffect`, but allows passing a value which, when it
+-- | changes, will trigger the effect to run again. E.g.:
+-- |
+-- | ```purs
+-- | view :: ReactElement
+-- | view = withHooks do
+-- |   count /\ setCount <- useState 0
+-- |
+-- |   useEffect' count \c -> liftEffect $
+-- |     HTMLDocument.setTitle ("You clicked " <> show c <> " times") =<< document =<< window
+-- |
+-- |   pure H.button_ "" { onClick: setCount $ count + 1 } "Click me"
+-- | ```
+useEffect' :: forall a. Eq a => a -> (a -> Aff Unit) -> Hook Unit
+useEffect' deps = \aff ->
+  useEffect_ name deps aff
+  where
+    name = uniqueNameFromCurrentCallStack { skipFrames: 3 }
 
--- | A version of `useEffect` that logs messages, state changes, render times,
--- | and info from the name-generating function. Intended to be used with
--- | qualified imports: `UseEffect.traced`.
+-- | A version of `useEffect` that logs info from the name-generating function.
+-- | Intended to be used with qualified imports: `UseEffect.traced`.
 traced :: DebugWarning => Aff Unit -> Hook Unit
-traced = useEffect' withTrace uniqueNameFromCurrentCallStackTraced
-
-useEffect' ::
-  (ComponentDef Void Unit -> ComponentDef Void Unit)
-  -> ({ skipFrames :: Int } -> ComponentName)
-  -> Aff Unit
-  -> Hook Unit
-useEffect' f genName init =
-  mkHook name \render -> f
-    { init: forkVoid init
-    , update: \_ msg -> absurd msg
+traced aff =
+  useEffect_ name unit $ const aff
+  where
+    name = uniqueNameFromCurrentCallStackTraced { skipFrames: 3 }
+
+-- | A version of `useEffect'` that logs info from the name-generating function.
+-- | Intended to be used with qualified imports: `UseEffect.traced'`.
+traced' :: forall a. DebugWarning => Eq a => a -> (a -> Aff Unit) -> Hook Unit
+traced' deps = \aff ->
+  useEffect_ name deps aff
+  where
+    name = uniqueNameFromCurrentCallStackTraced { skipFrames: 3 }
+
+useEffect_ :: forall a. Eq a => ComponentName -> a -> (a -> Aff Unit) -> Hook Unit
+useEffect_ name = \deps init ->
+  let _ = set name deps
+  in
+  mkHook name \render ->
+    { init: do
+        fork do
+          pure $ set name deps
+          init deps
+        pure deps
+    , update: \deps' _ -> do
+        let newDeps = get name
+        if deps' == newDeps then
+          fork $ pure unit
+        else
+          fork do
+            pure $ set name newDeps
+            init newDeps
+        pure newDeps
```

Comment:
This looks like an active wait to me. 100% CPU consumption. 
Or am I missing something again?

Reply:
Yeah, I guess this wasn’t great. Performance seemed okay — I thought it might be okay since it’s sort of asynchronous with the `delay`. But the `componentDidUpdate` idea is much better.